Adds the New-and-Improved Training Machine, Training Toolbox!

MrDoomBringer opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->
_This all started a few weeks ago when I hadn't played for like 6 months lost an honorable toolbox 1v1 against Propser Finely. It was at that point I realized I was sorely rusty, and required a way to rigorously train my clicking abilities. From that, this PR was born._

## About The Pull Request
From what I can tell the target stake in the firing range has literally not worked in years. On top of that, the code for it is kinda all over the place and not very user-friendly, so most people just ignore it. Exactly one map (Icebox station) even has it mapped correctly but still doesn't work there. 

So this PR replaces it with a whole new system:  **The Personnel Training Machine!**
The machine, when activated, will move around wildly in a circle. Has adjustable-but-constrained speed and range.
![image](https://user-images.githubusercontent.com/29008542/94839830-8b09c700-03e5-11eb-84a9-3ad6fa528c59.png)
look he even has a little chair to sit people down on

Also, I've added a special toolbox that does no damage, but instead counts the number of hits made against mobs and training machines. Annoy people and practice at the same time!

Both the toolbox and machine will be mapped into dorms. Replacements for both will be available from cargo.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
The Toolbox tourney is coming up. What better way to prepare than this?

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Nanotrasen has replaced the broken target stakes with a new model.
add: Training toolboxes have been added! Harmless, but useful for practicing your clicking skills and bugging your crewmembers at the same time.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

todo:

- [X] mapping in these things
- [X] cargo packs
- [X] some tgui work
- [X] robot rebellion when emagged
- [X] autodoc

MrDoomBringer commented:
As a note, this specific implementation (Notably using the /obj/structure/ type path for the training machine, and using timers instead of SSFastProcess) has received a degree of maintainer approval

Tlaltecuhtli commented:
make it fight back on emag

MrDoomBringer commented:
@Tlaltecuhtli that is an excellent idea

Rohesie commented on code/game/objects/items/shooting_range.dm:
```diff
@@ -4,41 +4,16 @@
 	..()
 	if(I.use_tool(src, user, 0, volume=40))
-		removeOverlays()
+		cut_overlays()
```
Can you only remove the relevant overlays here?

MrDoomBringer replied:
ok

Jared-Fogle commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,231 @@
+	var/total_hits = 0
+	var/lap_hits = 0
+
+/obj/item/training_toolbox/equipped(mob/user)
+	. = ..()
```
?

Jared-Fogle commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,231 @@
+	if (pinned_target)
+		. += "<span class='notice'><b>Alt-Click to remove pinned target</b></span>"
+	. += "<span class='notice'><b>Click to open control interface.</b>"
+/obj/item/training_toolbox
```
Leading new line

Jared-Fogle commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,231 @@
+		var/obj/structure/training_machine/trainer = target
+		if (!trainer.pinned_target)
+			return
+	total_hits ++
+	lap_hits ++
```
```suggestion
	total_hits++
	lap_hits++
```

Jared-Fogle commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,231 @@
+	lap_hits ++
+	user.changeNext_move(CLICK_CD_MELEE)
+	playsound(src,'sound/weapons/smash.ogg',50,FALSE)
+	if (lap_hits % 9 == 0)
```
Magic number, move to a define.

Jared-Fogle commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,231 @@
+
+/obj/structure/training_machine/attackby(obj/O, mob/user)
+	. = ..()
+	if (istype(O, /obj/item/target))
```
Early return. Also, better variable names. Preferably, just do `obj/item/target/target` then check `if (!istype(target))`.

Jared-Fogle commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,231 @@
+		. += "<span class='warning'>It has a dangerous-looking toolbox attached to it...</span>"
+	if (pinned_target)
+		. += "<span class='notice'><b>Alt-Click to remove pinned target</b></span>"
+	. += "<span class='notice'><b>Click to open control interface.</b>"
```
```suggestion
	. += "<span class='notice'><b>Click to open control interface.</b></span>"
```

MrDoomBringer commented:
@Tlaltecuhtli 
![image](https://user-images.githubusercontent.com/29008542/94862543-22cbdd00-0407-11eb-875d-e4e78c415204.png)
he's coming to fuck you up

ATH1909 commented:
add a tracker for how many times you accidentally hit yourself

also, does this move around in a fixed pattern, or does it move around semi-randomly?

ATH1909 commented:
it'd also be nice if it recorded ranged hits, in case you wanna practice your gunslinging with disablers or laser tag guns or something

Jared-Fogle commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -54,20 +54,21 @@
 	ui_interact(user)
 
-/obj/structure/training_machine/attackby(obj/O, mob/user)
+/obj/structure/training_machine/attackby(var/obj/item/target/target, mob/user)
```
```suggestion
/obj/structure/training_machine/attackby(obj/item/target/target, mob/user)
```

Jared-Fogle commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -2,7 +2,7 @@
 #define MAX_RANGE 7
 #define MAX_SPEED 10
-
+#define HITS_TO_KILL 9
```
```suggestion
#define HITS_TO_KILL 9

```

ATH1909 commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,234 @@
+		var/mob/living/carbon/target = locate() in view(1, src)
+		if (target?.stat == CONSCIOUS)
+			do_attack_animation(target, null, secret_evil_toolbox)
+			target.adjustBruteLoss(secret_evil_toolbox.force)
```
this ignores armor...

MrDoomBringer replied:
push it to the limit

MrDoomBringer replied:
9 hits to a drop everyone knows the rules

ATH1909 commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,234 @@
+		return
+	obj_flags |= EMAGGED
+	secret_evil_toolbox = new(src)
+	to_chat(user, "<span class='warning'>You override the training machine' safety protocols, and activate it's realistic combat feature. A toolbox pops out of a slot on the top.</span>")
```
```suggestion
	to_chat(user, "<span class='warning'>You override the training machine's safety protocols, and activate its realistic combat feature. A toolbox pops out of a slot on the top.</span>")
```

ATH1909 commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,234 @@
+	if (obj_flags & EMAGGED)
+		return
+	obj_flags |= EMAGGED
+	secret_evil_toolbox = new(src)
```
is this toolbox dropped when the emagged training machine is destroyed? if so, is that intentional?

MrDoomBringer replied:
maybe and if so yes
each one of these comes with a secret evil toolbox built in don't you know

MrDoomBringer replied:
ok fine

MrDoomBringer commented:
the todos done the reviews addressed now for the map conflicts to roll in

MrDoomBringer commented:
@ATH1909 the training machine itself technically doesn't count hits done - rather, the cool training toolbox does.

MrDoomBringer commented:
I could add a hits recieved counter to the training machine but that would involve a little bit of complexity - as we'd have to track hits received a buckled mob - so I'll leave that and other extra features for another PR after this one. For now, I'm just aiming to get this through the door so players can experiment with this before the tourney if they want to

Jared-Fogle commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -140,7 +142,7 @@
 			attack_cooldown = rand(MIN_ATTACK_DELAY,MAX_ATTACK_DELAY)
 
-	addtimer(CALLBACK(src, .proc/do_movement), MAX_SPEED - move_speed)
+	addtimer(CALLBACK(src, .proc/do_movement), abs(MAX_SPEED - move_speed) + 1) // We want to ensure this is never <=0
```
Are you sure this is what you want and not `max`? As in, `max(MAX_SPEED - move_speed, 1)`?

MrDoomBringer replied:
max works better for this u right

Jared-Fogle commented on code/game/objects/items/shooting_range.dm:
```diff
@@ -68,7 +68,7 @@
 			bullet_hole.icon_state = "dent"
 		bullethole_overlays += bullet_hole
-		add_overlay(bullet_hole)
+		add_overlay(buillet_hole)
```
?

MrDoomBringer replied:
very good mr doom bringer very good

MrDoomBringer committed:
first pass


training bot training bot


oops


coode


big iron
removals
tea gooey
more removals
code cleaning
remove only the relevant overlays is this how i do that
robot reveenge
now just to map and autodoc
Apply suggestions from code review

Co-authored-by: Jared-Fogle <35135081+Jared-Fogle@users.noreply.github.com>
Co-authored-by: ATH1909 <42606352+ATH1909@users.noreply.github.com>
autodoc
mapping it in
fix delta
booooring
might as well make it compile
tgui
update packs
pack 2
remove machine from packs, fix toolbox vs machine action
remove snowflake code
i freaked it

Rohesie commented on code/game/objects/items/shooting_range.dm:
```diff
@@ -4,41 +4,19 @@
-	. = ..()
-	if(pinnedLoc)
-		pinnedLoc.forceMove(loc)
+	var/list/bullethole_overlays = list()
```
Could you turn this into a lazy list?

Rohesie commented on code/game/objects/items/shooting_range.dm:
```diff
@@ -4,41 +4,19 @@
 	..()
 	if(I.use_tool(src, user, 0, volume=40))
-		removeOverlays()
+		for (var/image/bullethole in bullethole_overlays)
```
```suggestion
		for (var/bullethole in bullethole_overlays)
```
No need to type check here, right?

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,264 @@
+	if (pinned_target)
+		. += "<span class='notice'><b>Alt-Click to remove pinned target</b></span>"
+	. += "<span class='notice'><b>Click to open control interface.</b></span>"
+
+/obj/structure/training_machine/Destroy()
+	. = ..()
+	if (secret_evil_toolbox)
+		QDEL_NULL(secret_evil_toolbox)
```
Please put the `Initialize()` and `Destroy()` procs right after the object definition, for better organization.

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,264 @@
+	var/turf/starting_turf
+	///How fast the machine moves. Cannot be higher than MAX_SPEED
+	var/move_speed = 1
+	///Reference to a potentially attached obj/item/target 
+	var/obj/item/target/pinned_target
```
Could you clean this variable in `Destroy()` ?

stylemistake commented on tgui/packages/tgui/interfaces/TrainingMachine.js:
```diff
@@ -0,0 +1,74 @@
+                minValue={1}
+                maxValue={10}
+                onDrag={(e, value) => {
+                  act("movespeed", { movespeed: value });
```
why range is stored in local state, when it can be a part of ui_data?

MrDoomBringer replied:
for future reference, turns out the issue was me forgetting to return TRUE in ui_act()

MrDoomBringer committed:
tgui

MrDoomBringer committed:
Review addressing

MrDoomBringer committed:
early return

MrDoomBringer committed:
gets rid of localstate, other tweaks

MrDoomBringer committed:
bundle

MrDoomBringer committed:
lets make attached items generic

MrDoomBringer committed:
fight back

MrDoomBringer commented:
This should be ready for further-review/merge. 
Now you can attach a training toolbox to the machine and it'll fight you with it, and thrown training toolboxes counts hits. Emag behavior remains the same.

MrDoomBringer committed:
we can set this back to drop_location now

MrDoomBringer committed:
touch ups

Rohesie commented on code/game/objects/items/shooting_range.dm:
```diff
@@ -5,40 +5,21 @@
 	density = FALSE
 	var/hp = 1800
-	var/obj/structure/target_stake/pinnedLoc
+	//Lazylist to keep track of bullet-hole overlays
```
```suggestion
	/// Lazylist to keep track of bullet-hole overlays.
```

Rohesie commented on code/game/objects/items/shooting_range.dm:
```diff
@@ -5,40 +5,21 @@
-		removeOverlays()
+		for (var/bullethole in bullethole_overlays)
+			cut_overlay(bullethole)
+		LAZYCLEARLIST(bullethole_overlays)
```
```suggestion
		bullethole_overlays = null
```
This is a lazy list, we want it nulled if empty.

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+	return data
+
+/obj/structure/training_machine/ui_act(action, params)
+	if(..())
+		return
```
```suggestion
	. = ..()
	if(.)
		return
```
* Preserve the parent call returns.

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+	ui_interact(user)
+
+/obj/structure/training_machine/attackby(obj/item/target, mob/user)
+	. = ..()
+	if (user.a_intent == INTENT_HARM)
+		return
+	if (!istype(target, /obj/item/training_toolbox) && !istype(target, /obj/item/target))
+		return
```
You might want to either handle parent returning a value not to continue the attack or to only call parent when your desired interaction options are unavailable.

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+/obj/structure/training_machine/proc/remove_attached_item(mob/user, throwing = FALSE)
+	if (!attached_item)
+		return
+	vis_contents.Cut()
```
```suggestion
	vis_contents -= attached_item
```

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+	to_chat(user, "<span class='notice'>You remove \the [attached_item] from the training device.</span>")
+	remove_attached_item(user)
+	playsound(src, "rustle", 50, TRUE)
+
+/obj/structure/training_machine/proc/toggle()
```
Could you document the new procs?

MrDoomBringer replied:
Sure

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+	if (!Move(nextStep, get_dir(src, nextStep)))
+		target_position = null //We couldn't move towards the target turf, so find a new target turf
+	try_attack()
+	addtimer(CALLBACK(src, .proc/do_movement), max(MAX_SPEED - move_speed, 1)) // We want to ensure this is never <=0
```
I've made use and abuse of timers for my autofire component on tgmc, but then ninjanomnom made a good point in that this overburdens the timer SS, too many things that are not one-time events relying on it.
Could you turn this into a process instead?

MrDoomBringer replied:
Is it possible to make the process have a dynamic wait time? The movespeed for this needs to be fairly granular/variable, and after speaking to a maintainer about it they agreed timers might be the best solution here. I might be missing something though

Rohesie replied:
Explained how to handle it via discord.
What maintainer was that?

MrDoomBringer replied:
@Rohesie @SpaceManiac

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+	var/move_speed = 1
+	///Reference to a potentially attached object, either a target, trainer toolbox, or syndicate toolbox
+	var/obj/item/attached_item
+	///Time between attacks when emagged
+	var/attack_cooldown = 50
```
```suggestion
	var/attack_cooldown = 5 SECONDS
```
* Use time defines.

Rohesie replied:
```suggestion
```
This could be removed, with usage of the cooldown defines, since it won't be var-editable anyway, as it's getting reset on every attack.

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+	///Time between attacks when emagged
+	var/attack_cooldown = 50
+	///Helper for timing attacks when emagged
+	var/last_attack_time = 0
```
Could you use the macros defined in `cooldowns.dm` ?

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+/obj/structure/training_machine/proc/try_attack()
+	if (!attached_item || istype(attached_item, /obj/item/target))
+		return
+	if (world.time < last_attack_time + attack_cooldown)
```
```suggestion
	if (!COOLDOWN_FINISHED(src, attack_cooldown))
```

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+	if (obj_flags & EMAGGED)
+		target.apply_damage(attached_item.force, BRUTE, BODY_ZONE_CHEST)
+	playsound(src, 'sound/weapons/smash.ogg', 25, TRUE)
+	last_attack_time = world.time
```
```suggestion
	COOLDOWN_START(src, attack_cooldown, attack_cooldown_time)
```

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+		target.apply_damage(attached_item.force, BRUTE, BODY_ZONE_CHEST)
+	playsound(src, 'sound/weapons/smash.ogg', 25, TRUE)
+	last_attack_time = world.time
+	attack_cooldown = rand(MIN_ATTACK_DELAY, MAX_ATTACK_DELAY)
```
```suggestion
	attack_cooldown_time = rand(MIN_ATTACK_DELAY, MAX_ATTACK_DELAY)
```

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+	if (world.time < last_attack_time + attack_cooldown)
+		return
+	var/list/targets
+	for(var/mob/living/carbon/target in oview(1, get_turf(src))) //Find adjacent target
```
```suggestion
	for(var/mob/living/carbon/target in view(1, get_turf(src))) //Find adjacent target
```
Do you want to ignore targets which might happen to be on the same tile (teleported, getting there before the mahine became dense, or who knows how else)?

MrDoomBringer replied:
Yes, otherwise it will hit mobs buckled to it. I want to train against a monkey buckled to a target stake, after all.
Also, what would be the best way to get a list of mobs adjacent to this, if not oview()?

Rohesie replied:
Ah, then it's fine.

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+
+/obj/structure/training_machine/Destroy()
+	remove_attached_item(throwing = TRUE)
+	explosion(src, 0,0,1, flame_range = 2)
```
Can you move this to one of the object damage procs, such as `obj_destruction`? This would produce weird interactions on admin deletes, and explosions on qdeleted objects is not very kosher.

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+	to_chat(user, "<span class='notice'>You attach \the [attached_item] to the training device.</span>")
+	playsound(src, "rustle", 50, TRUE)
+
+/obj/structure/training_machine/proc/attach_item(target)
+	remove_attached_item()
+	attached_item = target
+	attached_item.forceMove(src)
+	attached_item.vis_flags |= VIS_INHERIT_ID
+	vis_contents += attached_item
```
Could you register to the `attached_item`'s `COMSIG_PARENT_QDELETING` event and handle it properly?

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -0,0 +1,310 @@
+	return TRUE
+
+/obj/item/training_toolbox/throw_impact(atom/hit_atom, datum/thrownthing/throwingdatum)
+	..()
```
```suggestion
	. = ..()
```

MrDoomBringer replied:
Actually, wouldn't it be better to have
```
if (!..())
  check_hit(hit_atom)
```
so things like getting caught by another mob won't trigger it?

Rohesie replied:
If you do that then use this pattern:
```cpp
. = ..()
if(!.)
```
Always preserve the parent call's return unless you have a very good reason not to.

MrDoomBringer committed:
Apply suggestions from Rohesie's code review

Co-authored-by: Rohesie <rohesie@gmail.com>

MrDoomBringer committed:
Attack Cooldown from Rohsie

Co-authored-by: Rohesie <rohesie@gmail.com>

MrDoomBringer committed:
attackby returns

MrDoomBringer committed:
Merge remote-tracking branch 'doombringer/toolbox_bot' into toolbox_bot

Rohesie commented:
You don't generally need to autodoc every proc, just the new ones you add. The redefined ones inherit the documentation at the base.

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -94,14 +120,43 @@
+ * contents to somehow be deleted.
+ */
+/obj/structure/training_machine/proc/on_attached_delete()
+	message_admins("On attach delete called")
```
Debug info?

MrDoomBringer replied:
Yes, sorry - didn't push the last commit

MrDoomBringer commented:
@Rohesie I know, but I felt that documentation could help for some of the overridden procs in any case.

Rohesie commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -94,14 +120,43 @@
+ */
+/obj/structure/training_machine/proc/on_attached_delete()
+	message_admins("On attach delete called")
+	vis_contents -= attached_item
+	attached_item = null
+	UnregisterSignal(attached_item, COMSIG_PARENT_QDELETING)
```
I'd put these into a proc or macro and call it by both `remove_attached_item()` and this proc.
Because there is behavior you are likely to miss adding to one that would make sense to be on the other, like `handle_density()`

Rohesie replied:
This is still valid, despite the original code is outdated.

MrDoomBringer replied:
Went ahead and made remove_attached_item() call this proc

MrDoomBringer committed:
using cooldown macros with ssfastprocess worked surprisingly well
autodoc
dimminsdale dimmaformatting
intent tweak
further tweaks
calling on_attached_delete

MrDoomBringer committed:
Merge remote-tracking branch 'origin/master' into toolbox_bot

MrDoomBringer committed:
tea gooey

MrDoomBringer commented:
Just gave another round of testing, no runtimes from what I can tell.
